src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php

```diff
@@ -110,15 +112,16 @@ protected function extractAttributes($object, $format = null, array $context = a
      */
     protected function getAttributeValue($object, $attribute, $format = null, array $context = array())
     {
-        if (null !== $this->classDiscriminatorResolver) {
-            $mapping = $this->classDiscriminatorResolver->getMappingForMappedObject($object);
-
-            if (null !== $mapping && $attribute == $mapping->getTypeProperty()) {
-                return $this->classDiscriminatorResolver->getTypeForMappedObject($object);
+        $cacheKey = spl_object_hash($object);
```

Comment:
I don't think this works:
- this creates a memory leak since the cache array is never cleared
- but more importantly, object hashes can be reused so that a cached value can map to another object

Both issues could be fixed by clearing this cache in the appropriate place (when exiting any recursive normalization steps)
On PHP 7.2, we should use spl_object_id btw.

Reply:
>Both issues could be fixed by clearing this cache in the appropriate place (when exiting any recursive normalization steps)

So in the `Serializer` class. A bit intrusive. (we'll have to introduce a new interface, and call a `clearCache()` method if the normalizer implements this interface.

Another solution (cleaner) would be to do the resolving in the method calling `getAttributeValue()`, and pass the result as a new argument of getAttributeValue().
Because it's a protected method, will have to deprecate not passing this new argument.

Reply:
`ResetInterface` doesn't work as explained: handles can be reused between calls.
The `spl_object_id` polyfill adds overhead. Since we're tracking perf here, better not.
I'll review iteration v3 :)

Reply:
If using the polyfill, this PR would still be faster than the current 4.1 state (I'm trying to profile using php 7.1, but I got weird results because of `DateTimeZone::getTransitions`: https://blackfire.io/profiles/6c0105f5-bf95-49d3-a49d-25adc16cd6fc/graph)!

Reply:
@dunglas what about going back to v1 slightly improved to store the discriminator attribute per class and call `ClassDiscriminatorResolverInterface::getMappingForMappedObject` only when `getAttributeValue` is called when `$attribute === $this->discriminatorCache[\get_class($object)]`?

This would avoid using `spl_object_id`/`spl_object_hash` and the cache will have a finite size and will be reusable between requests.

This would improve perfs a lot, because, we are only computing the cache value once per class (instead of object) and `ClassDiscriminatorResolverInterface::getMappingForMappedObject` will be called only once per object which is involved in a class discriminator.

I will give this a try and call it v4 ;)